File: src/com/kraftwerking/datastructuresalgo/leetcode/GraphValidTree261.java
package com.kraftwerking.datastructuresalgo.leetcode;


import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

class GraphValidTree261 {

  Set<Integer> visit = new HashSet<>();
  Map<Integer, List<Integer>> adj = new HashMap<>();

  public boolean validTree(int n, int[][] edges) {
    if (n < 1) {
      return true;
    }
    for (int i = 0; i < n; i++) {
      adj.put(i, new ArrayList<>());
    }
    for (int[] edge : edges) {
      adj.get(edge[0]).add(edge[1]);
      adj.get(edge[1]).add(edge[0]);
    }

    return dfs(0, -1) && n == visit.size();
  }

  public boolean dfs(int i, int prev) {
    if (visit.contains(i)) {
      return false;
    }
    visit.add(i);
    for (int j : adj.get(i)) {
      if (j == prev) {
        continue;
      }
      if (!dfs(j, i)) {
        return false;
      }
    }
    return true;
  }

  public static void main(String[] args) {
    int n = 5;
    int[][] edges = {{0, 1}, {0, 2}, {0, 3}, {1, 4}};

    GraphValidTree261 graphValidTree261 = new GraphValidTree261();
    boolean result = graphValidTree261.validTree(n, edges);
    System.out.println(result);
  }
}

File: python/GraphValidTree261.py
from typing import List


class Solution(object):
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        if not n:
            return True

        adj = {i: [] for i in range(n)}
        for n1, n2 in edges:
            adj[n1].append(n2)
            adj[n2].append(n1)

        visit = set()

        def dfs(i, prev):
            if i in visit:
                return False

            visit.add(i)
            for j in adj[i]:
                if j == prev:
                    continue
                if not dfs(j, i):
                    return False
            return True

        return dfs(0, -1) and n == len(visit)

        return True


def main():
    n = 5
    edges = [[0, 1], [0, 2], [0, 3], [1, 4]]

    solution = Solution()
    result = solution.validTree(n, edges)
    print(result)


if __name__ == "__main__":
    main()
